docs/ApiGuidelines.md

```diff
@@ -0,0 +1,76 @@
+# API Guidelines
+
+## APIs
+
+There are three types of APIs: 
+
+1. Public
+2. Experimental
+3. Internal
+
+### Public APIs
+
+Public APIs are the supported APIs that apps can use. 
+
+Public API requirements: 
+
+* All public APIs must have at least one test case.
+* All public APIs must be in a header file that is published to a location under the build directory.
+* All public APIs must be fully documented using doxygen markup in the public header files
+* All public APIs must be used in some sample code or snippet. Ideally, the sample
+  should be one that is compiled as part of the build, and uses no experimental APIs. For rarely 
+  used public APIs, the sample might simply be a snippet in the doxygen comments.
+* Public APIs cannot have source-breaking changes across Open Enclave versions. For platforms 
+  where shared libraries are build, public APIs cannot have binary-breaking changes across 
+  Open Enclave versions either. (Source- and binary-breaking changes can still be made to a new 
+  API before it is released for the first time.)
```

Comment:
When you say versions, do you mean all versions, major revisions? Committing to eternal binary compatibility for .so sounds very onerous, I think at least major releases should allow this to be revisited (likewise for source compatibility, really).

Reply:
Nevermind, I should have read the bit about Breaking below. It might make sense to mention the exception here as well, in case someone reads just this bullet point and gets the wrong impression.

Reply:
Added forward reference